server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/vacation/MemoryMailboxChangeRepository.java

```diff
@@ -87,39 +88,50 @@ public MailboxChangeCollector(int limit) {
             }
 
             private State state;
+            private ZonedDateTime date;
             private boolean hasMoreChanges;
-            private int numberOfChanges;
-            private int limit;
-            private List<MailboxId> created;
-            private List<MailboxId> updated;
-            private List<MailboxId> destroyed;
+            private Limit limit;
+            private Set<MailboxId> created;
+            private Set<MailboxId> updated;
+            private Set<MailboxId> destroyed;
 
-            public MailboxChangesBuilder(int limit) {
+            public MailboxChangesBuilder(Limit limit) {
                 this.limit = limit;
-                this.created = new ArrayList<>();
-                this.updated = new ArrayList<>();
-                this.destroyed = new ArrayList<>();
+                this.hasMoreChanges = false;
+                this.created = new HashSet<>();
+                this.updated = new HashSet<>();
+                this.destroyed = new HashSet<>();
             }
 
             public MailboxChangesBuilder add(MailboxChange change) {
-                if (numberOfChanges + change.getTotalChanges() > limit) {
+                Set<MailboxId> createdTemp = new HashSet<>(created);
+                Set<MailboxId> updatedTemp = new HashSet<>(updated);
+                Set<MailboxId> destroyedTemp = new HashSet<>(destroyed);
+                createdTemp.addAll(change.getCreated());
+                updatedTemp.addAll(change.getUpdated());
+                destroyedTemp.addAll(change.getDestroyed());
+
+                if (createdTemp.size() + updatedTemp.size() + destroyedTemp.size() > limit.getValue()) {
                     hasMoreChanges = true;
                     return this;
                 }
 
+                date = change.getDate();
                 state = change.getState();
                 this.created.addAll(change.getCreated());
                 this.updated.addAll(change.getUpdated());
                 this.destroyed.addAll(change.getDestroyed());
-                this.numberOfChanges += change.getCreated().size() + change.getUpdated().size() + change.getDestroyed().size();
 
                 return this;
             }
 
             public MailboxChangesBuilder addAll(MailboxChangesBuilder builder) {
-                this.created.addAll(builder.created);
-                this.updated.addAll(builder.updated);
-                this.destroyed.addAll(builder.destroyed);
+                if (created.size() + updated.size() + destroyed.size() + builder.created.size() + builder.updated.size() + destroyed.size() > limit.getValue()) {
```

Comment:
Extract `created.size() + updated.size() + destroyed.size()` in a submethod to not repeat yourself

Reply:
This is the only occurrence?